src/mqtt_client/mqtt_client.rs

```diff
@@ -321,14 +336,28 @@ where
                 }
             };
 
-            log::info!("Settings update: `{}`", path);
-
-            let message: SettingsResponse = settings
-                .string_set(path.split('/').peekable(), message)
-                .map(|_| {
-                    update = true;
-                })
-                .into();
+            let mut buffer = [0; MESSAGE_SIZE];
+            let message: SettingsResponse =
+                match settings.string_get(path.split('/').peekable(), &mut buffer) {
+                    Ok(len) => {
+                        match settings.string_set(path.split('/').peekable(), message) {
+                            Ok(_) => {
+                                updated = true;
+                                let result = handler(settings);
+                                if result.is_err() {
+                                    // Note(unwrap): We just serialized this value, so it should
+                                    // always be valid.
+                                    settings
+                                        .string_set(path.split('/').peekable(), &buffer[..len])
+                                        .unwrap();
```

Comment:
Need to set updated to false if we rejected the settings.

I think we also need to republish the correct value.